consensus/istanbul/backend/announce.md

```diff
@@ -0,0 +1,140 @@
+
+It is important to note that a validator's `eNodeURL` is never shared among third parties, but directly published to specific peers. Therefore it is allowed for a validator to publish different external facing `eNodeURL` values to different validator peers.
+
+## Basic Operation
+
+When a validator is close to being elected ([NearlyElectedValidator]), it starts periodically sending [queryEnodeMsg] messages through the p2p network. These message are regossipped by nodes, and validators reply with a direct message to the originator with an [eNodeCertificateMsg], holding their `eNodeURL`. The initial [queryEnodeMsg] contained the origin validator's `eNodeURL` encrypted with the destination public key, therefore after the direct reply, both validators are aware of each others' `eNodeURL`.
+
+### Keeping eNodeURL's up to date
+
+To ensure that validator `eNodeURL` tables don't get stale, each validator's `eNodeURL` is paired with a version number, which is both attached to the origin [queryEnodeMsg], and in the [eNodeCertificateMsg] reply. Validators should ignore enodeurl from versions older than the highest known one. The current convention used is that versions are unix timestamps from the moment of the update.
+
+### Minimizing network traffic
+
+In order to reduce the amount of message flooding, there's a third type of message: the [versionCertificateMsg]. Periodically (or when it's updated by the validator operators) a validator will update its version, and gossip through the network a VersionCertificateMsg holding only its own new version.
+
+When a node (every full node in the network) receives a [versionCertificateMsg], it will update the highest known versions according to the certificates, and regossip a VersionCertificateMsg holding only the version certificates that were new to it.
+
+This allows every validator to have an idea of what the highest known version is for every other validator's `eNodeURL`, and in the next query only request for those that are stale.
+
+### Proxy agnosticism
+
+The announce protocol itself is agnostic to the concept of proxies, it cares only for the sharing of `<validator address, validator eNodeUrl>` tuples. It is the proxy implementation's responsibility to ensure the correct behavior of this specification.
+
+### p2p connection management
+
+Wether a validator opens a connection or not against another validator is not in the scope of this spec, but only how and when should the protocol messages should be sent, replied and regossipped. For example, it IS part of the spec what messages should be sent to a newly connected peer or validator peer.
+
+## Protocol Messages (RLP)
+
+// add types with the RLP nomenclature
+
+### queryEnodeMsg (0x12)
+
+`[[dest_address: B_20, encrypted_enode_url: B], version: P, timestamp: P]`
```

Comment:
It would be good to add a reference explaining the meaning of B_20, P ... etc

Reply:
I think this message is missing the validator celo address

Reply:
Isn't `dest_address` the 20 byte celo address of the destination validator?

Agree on the types, I'll try to add a link to the RLP nomenclature

Reply:
So `dest_address` is the address of the validator that the enode is encrypted for but the message is lacking the source address (which is contained in the istanbul message), which the receiver will require to be able to map validator address -> enode.

Reply:
Ah, I misunderstood you before. Ok, you mean the origin address. You are correct. I will clarify that.